edge-hub/src/Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter/ISubscriptionWatcher.cs

```diff
@@ -0,0 +1,10 @@
+// Copyright (c) Microsoft. All rights reserved.
+namespace Microsoft.Azure.Devices.Edge.Hub.MqttBrokerAdapter
+{
+    using System.Collections.Generic;
+
+    public interface ISubscriptionWatcher
+    {
+        IReadOnlyCollection<SubscriptionPattern> WatchedSubscriptions { get; }
```

Comment:
2 questions - 
1. Should the consumer/subscriber/SubscriptionWatcher be the same interface?
2. If we move all the subscription matching to one class as I suggested above, then we don't really need this?

Reply:
The Consumer/Subscriber could be merged as today they go in pairs (all classes implementing 'subscriber' also implements 'consumer' and it consumes the topic 'subscriber' subscribed to)
SubscriptionWatcher is a different thing and it is not related to 'Subscriber' even if the name implies: in case of a 'Subscriber', the implementor class has needs to see a specific topic (so then later it can receive those messages). In case of SubscriptionWatcher, the implementor class needs to no if a connected client subscribed to a specific topic. 
Basically it is the opposite direction: if I am a subscriber, I need the messages. If I am a SubscriptionWatcher, I watch who subscribed to a topic, so I can provide them messages.

Reply:
Refactored this the following way:
- ISubscriber/IMessageConsumer is merged at a price that SubscriptionChangeHandler cannot use ComponentDiscovery to find the classes (that is to eliminate circular dependencies), but it gets all the interested components injected directly in its constructor
- Removed the interface ISubscriptionWatcher at the price that its single method is added to the interface of all three original implementor.

Now
- Have only two interfaces from the original four (IMessageConsumer/Producer)
- The main loop in the connector is simpler (it does not have SubscriptionChangeHandler as a special case, but it works as a normal MessageConsumer)

But
- the SubscriptionChangeHandler cannot pick up automatically the classes that needs subscription changes to be handled, but needs to add to the parameter list of SubscriptionChangeHandler. More over, that class does not need the instances for other reason just to get the matching patterns. It could have receive a list, but that requires either keyed parameters (-> adding dependency to autofac), or manual registration (-> does not work with ComponentDiscovery)